crates/ruma-server-util/src/authorization.rs

```diff
@@ -37,204 +43,164 @@ impl XMatrix {
+                    } else {
+                        origin = Some(OwnedServerName::try_from(value.to_unescaped())?);
                     }
-                    origin = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "destination" => {
                     if destination.is_some() {
-                        debug!("Field destination duplicated in X-Matrix Authorization header");
+                        debug!("destination parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        destination = Some(OwnedServerName::try_from(value.to_unescaped())?);
                     }
-                    destination = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "key" => {
                     if key.is_some() {
-                        debug!("Field key duplicated in X-Matrix Authorization header");
+                        debug!("key parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        key = Some(OwnedServerSigningKeyId::try_from(value.to_unescaped())?);
                     }
-                    key = Some(std::str::from_utf8(&value).ok()?.try_into().ok()?);
                 }
                 "sig" => {
                     if sig.is_some() {
-                        debug!("Field sig duplicated in X-Matrix Authorization header");
+                        debug!("key parameter duplicated in X-Matrix Authorization header");
+                    } else {
+                        sig = Some(value.to_unescaped());
                     }
-                    sig = Some(std::str::from_utf8(&value).ok()?.to_owned());
                 }
-                name => {
-                    debug!("Unknown field {} found in X-Matrix Authorization header", name);
+                _ => {
+                    debug!("Unknown parameter {name} in X-Matrix Authorization header");
                 }
             }
         }
 
-        Some(XMatrix { origin: origin?, destination, key: key?, sig: sig? })
-    })
+        Ok(Self {
+            origin: origin
+                .ok_or_else(|| XMatrixParseError::MissingParameter("origin".to_owned()))?,
+            destination,
+            key: key.ok_or_else(|| XMatrixParseError::MissingParameter("key".to_owned()))?,
+            sig: sig.ok_or_else(|| XMatrixParseError::MissingParameter("sig".to_owned()))?,
+        })
+    }
 }
 
-fn is_alpha(c: u8) -> bool {
-    (0x41..=0x5A).contains(&c) || (0x61..=0x7A).contains(&c)
+impl fmt::Debug for XMatrix {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("XMatrix")
+            .field("origin", &self.origin)
+            .field("destination", &self.destination)
+            .field("key", &self.key)
+            .finish_non_exhaustive()
+    }
```

Comment:
Why do we not show `sig`?

Reply:
Because it's an opaque string and I don't think it has much value in logs? I could leave it if you think it makes more sense.

Reply:
Hmm, I guess I am agnostic then, since it likely has no value, but it could be annoying to not have if you need it to debug something. :shrug:

Reply:
Sounds good!